src/Microsoft.Graph.Core/Authentication/TokenCredentials/IntegratedWindowsTokenCredential.cs

```diff
@@ -0,0 +1,138 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph
+{
+    using Azure.Core;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Collections.Generic;
+    using Microsoft.Identity.Client;
+    using System;
+
+    /// <summary>
+    /// A <see cref="TokenCredential"/> implementation using MSAL.Net to acquire token by integrated windows authentication.
+    /// </summary>
+    public class IntegratedWindowsTokenCredential : TokenCredential
+    {
+
+        private static readonly List<string> WellKnownTenants = new List<string>
```

Comment:
```suggestion
        private static readonly HashSet<string> WellKnownTenants = new HashSet<string>
```

and replace any string lookup, complexity becomes o(1) instead of o(n)